### PR TITLE
Added scheduler plugin functionality

### PR DESCRIPTION
**Pull request by tsydd**
Adds ability to customize scheduler logic using external plugins

### CODE REVIEW

**Comment by jdef** [comment]
Thanks for the PR! A few points:

1. We **strongly** prefer to accept PR's against master and backport, as needed, to prior releases. This PR is against releases/1.3, not master. Please retrofit this code and open a PR against master instead. If it lands there, then we'll consider backporting it to prior releases.
1. Nice example plugin! We have a public repo for those though (https://github.com/mesosphere/marathon-example-plugins), which is a better fit than the `plugin-interface` project, which we generally reserve for interfaces only. We're cleaning up the existing examples too in an effort to improve things for the community. Your example would do well here.
1. The integration of this plugin into the main scheduling logic appears to "further refine" existing scheduler decisions (vs. replacing them alltogether); documenting this nuance is likely important.
1. Guice is the bane of our existence. Seriously. Can we find a way to lessen the integration with it?

**Comment by tsydd** [comment]
Rebased onto master. I have no idea how to get rid of guice when integrating PluginManager into ResourceMatcher

File: src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginConfigurer.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  * @author Dmitry Tsydzik.
+  * @since 6/15/17.
+  */
+class SchedulerPluginConfigurer @Inject() (pluginManager: PluginManager) {
+  private[this] val log = LoggerFactory.getLogger(getClass)
+
+  try {
+    SchedulerPluginRunner.plugins = pluginManager.plugins[SchedulerPlugin].toList
+  } catch {
+    case e: Throwable =>
```
Comment (jdef) [inline src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginConfigurer.scala]: prefer `NonFatal` to `Throwable`

File: src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginConfigurer.scala [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+  * Workaround to integrate [[mesosphere.marathon.core.scheduler.SchedulerPluginRunner]] (object)
+  * with [[mesosphere.marathon.core.plugin.PluginManager]] (created with Guice)
+  *
+  * @author Dmitry Tsydzik.
+  * @since 6/15/17.
+  */
+class SchedulerPluginConfigurer @Inject() (pluginManager: PluginManager) {
+  private[this] val log = LoggerFactory.getLogger(getClass)
```
Comment (jdef) [inline src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginConfigurer.scala]: prefer `StrictLogging`

File: src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginConfigurer.scala [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+  try {
+    SchedulerPluginRunner.plugins = pluginManager.plugins[SchedulerPlugin].toList
+  } catch {
+    case e: Throwable =>
+      log.error("Failed to load plugins", e)
```
Comment (jdef) [inline src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginConfigurer.scala]: is it really worth using `error` level logging if the plan is to rethrow the exception? maybe `debug`  or `info` level would be more appropriate?

File: src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginRunner.scala [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  * Runs external scheduler plugins
+  *
+  * @author Dmitry Tsydzik.
+  * @since 6/15/17.
+  */
+object SchedulerPluginRunner {
+  /**
+    * Is populated by [[mesosphere.marathon.core.scheduler.SchedulerPluginConfigurer]]
+    */
+  var plugins: List[SchedulerPlugin] = List.empty
```
Comment (jdef) [inline src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginRunner.scala]: dislike global mutable state. let's avoid this

Comment (tsydd) [inline src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginRunner.scala]: I also dislike this approach. But I don't see another way to get list of plugins here. Can you suggest how to fix it?

File: src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginConfigurer.scala (hunk 2 above)
Comment (tsydd) [inline src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginConfigurer.scala]: fixed

File: src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginConfigurer.scala (hunk 1 above)
Comment (tsydd) [inline src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginConfigurer.scala]: fixed

File: src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginConfigurer.scala (hunk 3 above)
Comment (tsydd) [inline src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginConfigurer.scala]: fixed

**Comment by brat002** [comment]
Does any chance to merge this PR in master soon?

File: src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginRunner.scala (hunk 4 above)
Comment (unterstein) [inline src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginRunner.scala]: Would it be possible to introduce a SchedulerPluginModule where we inject the PluginsManager and hold the list of SchedulerPlugins and wire this to the needed classes?

**Comment by unterstein** [comment]
hey @tsydd,

still on it? Do you need assistance or a sparring partner for the `var plugins: List[SchedulerPlugin]` discussion?

Thanks
Johannes

**Comment by tsydd** [comment]
Updated plugin.md

File: src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginRunner.scala (hunk 4 above)
Comment (tsydd) [inline src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginRunner.scala]: It doesn't solve the issue. The issue is that we have to integrate different scopes: guice (pluginManager) and global (ResourceMatcher).

Comment (unterstein) [inline src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginRunner.scala]: `ResourceMatcher.matchResources(..)` is called by `InstanceOpFactoryImpl` where we do have the plugin manager. At this point we could either pass a list of `SchedulerPlugin` or you could hide this list behind trait abstracting this list. Would this be an option?

Comment (tsydd) [reply src/main/scala/mesosphere/marathon/core/scheduler/SchedulerPluginRunner.scala]: fixed

**Comment by tsydd** [comment]
I've fixed pr according to your comments. Can it be merged now?

**Comment by unterstein** [comment]
Hey @jdef, could you please have a look at the addressed changes? This would be awesome :)

**Comment by jeschkies** [comment]
Shouldn't this run through the CI once?

**Comment by jdef** [comment]
yes @jeschkies it should definitely run through CI

**Comment by unterstein** [comment]
wondering why we don`t have CI for community contributions. Anyways, I am currently setting up a phabricator review just for running CI and if its done, I`ll post the results here and merge if everything is fine 🎉

**Comment by unterstein** [comment]
Phab review here:
https://phabricator.mesosphere.com/D1065

Jenkins job here:
https://jenkins.mesosphere.com/service/jenkins/job/public-marathon-phabricator/1024/

**Comment by unterstein** [comment]
CI passed, happy to ship this commit!

Thank you very much @tsydd!

```
    .  o ..                  
     o . o o.o                
          ...oo               
            __[]__          Shippin'  
         __|_o_o_o\__         it!
         \""""""""""/         
          \. ..  . /          
     ^^^^^^^^^^^^^^^^^^^^
```

**Comment by brat002** [comment]
Thanks all. Great work!

**Pull request merged**